src/metrics/family.rs

```diff
@@ -222,10 +222,10 @@ impl<S: Clone + std::hash::Hash + Eq, M, C: MetricConstructor<M>> Family<S, M, C
             return metric;
         }
 
-        let mut write_guard = self.metrics.write();
-        write_guard.insert(label_set.clone(), self.constructor.new_metric());
-
-        drop(write_guard);
+        self.metrics
+            .write()
+            .entry(label_set.clone())
+            .or_insert_with(|| self.constructor.new_metric());
 
         RwLockReadGuard::map(self.metrics.read(), |metrics| {
```

Comment:
With `Family::remove` there is actually a second race condition.

In the scenarios where we have two threads:

A: Call `Family::get_or_create` with a new label set.
    1. Acquiring the read lock. Nothing there.
    2. Acquiring the write lock. Inserting the new metric.
    3. Being preempted by the OS.
B: Call `Family::remove` with the above label set.
A: Acquiring the read lock for the second (last) time, `.expect()`ing the previously inserted metric to be there.

This can be prevented by leveraging `RwLockWriteGuard::downgrade` of the previously acquired write lock instead of doing acquiring another read lock via `self.metrics.read`.

Does the above make sense?

Reply:
It makes sense yes. The patch now uses `RwLockWriteGuard::downgrade`.